File: ai-service/app/services/agentic_service.py
from datetime import datetime
from fastapi import HTTPException
from app.interfaces.i_agentic_service import IAgenticService
from app.models.schemas import ApprovalQueueItem, AgentActionSuggestion


class AgenticServiceImpl(IAgenticService):
    """
    Human-in-the-loop approval queue for agentic actions.
    SRP: manages the lifecycle of proposed agentic actions only.
    No AI logic lives here — that belongs in agents/.
    """

    def __init__(self) -> None:
        self._queue: dict[str, ApprovalQueueItem] = {}
        self._seed_demo_items()

    def propose_action(self, suggestion: AgentActionSuggestion, context: dict) -> ApprovalQueueItem:
        item = ApprovalQueueItem(
            proposed_action=suggestion.action,
            description=suggestion.description,
            impact=suggestion.impact,
            risk_level=suggestion.risk_level,
            tier=suggestion.tier,
            context=context,
        )
        self._queue[item.id] = item
        return item

    def get_queue(self) -> list[ApprovalQueueItem]:
        return sorted(self._queue.values(), key=lambda x: x.created_at, reverse=True)

    def approve(self, item_id: str, approved_by: str = "system") -> ApprovalQueueItem:
        item = self._get_or_raise(item_id)
        if item.status != "pending":
            raise HTTPException(status_code=400, detail=f"Item is already {item.status}")
        item.status = "approved"
        item.resolved_by = approved_by
        item.resolved_at = datetime.utcnow().isoformat()
        return item

    def reject(self, item_id: str, reason: str = "") -> ApprovalQueueItem:
        item = self._get_or_raise(item_id)
        if item.status != "pending":
            raise HTTPException(status_code=400, detail=f"Item is already {item.status}")
        item.status = "rejected"
        item.resolved_at = datetime.utcnow().isoformat()
        item.result = reason or "Rejected by reviewer"
        return item

    async def execute(self, item_id: str) -> ApprovalQueueItem:
        item = self._get_or_raise(item_id)
        if item.status != "approved":
            raise HTTPException(status_code=400, detail="Only approved actions can be executed")
        item.status = "executed"
        item.result = f"Action '{item.proposed_action}' executed successfully at {datetime.utcnow().isoformat()}"
        return item

    def _get_or_raise(self, item_id: str) -> ApprovalQueueItem:
        item = self._queue.get(item_id)
        if not item:
            raise HTTPException(status_code=404, detail=f"Approval item '{item_id}' not found")
        return item

    def _seed_demo_items(self) -> None:
        demos = [
            ApprovalQueueItem(
                proposed_action="schedule-attestation",
                description="Schedule quarterly control attestation for PAM domain",
                impact="Attestation requests dispatched to 4 control owners",
                risk_level="medium",
                tier="executor",
                context={"domain": "Access Management", "controls": 4},
            ),
            ApprovalQueueItem(
                proposed_action="create-workflow",
                description="Initiate governance workflow for Shadow AI risk issue",
                impact="Creates tracked remediation workflow with CISO approval stage",
                risk_level="high",
                tier="analyst",
                context={"risk": "Shadow AI", "priority": "P1"},
            ),
        ]
        for d in demos:
            self._queue[d.id] = d


agentic_service: IAgenticService = AgenticServiceImpl()

File: server/src/ai/AgenticService.ts
import { v4 as uuidv4 } from 'uuid';
import type { AgentAction, ApprovalQueueItem, ApprovalStatus, AgentTier } from './types';
import { QueueEngine } from '../queue/QueueEngine';

export interface IAgenticService {
  proposeAction(label: string, description: string, tier: AgentTier, payload: unknown): AgentAction;
  getApprovalQueue(): ApprovalQueueItem[];
  approve(id: string, decidedBy: string, note?: string): ApprovalQueueItem;
  reject(id: string, decidedBy: string, note?: string): ApprovalQueueItem;
  executeApproved(id: string): Promise<ApprovalQueueItem>;
}

class AgenticServiceImpl implements IAgenticService {
  private readonly queue = new Map<string, ApprovalQueueItem>();

  proposeAction(label: string, description: string, tier: AgentTier, payload: unknown): AgentAction {
    const action: AgentAction = {
      id:          uuidv4(),
      label,
      description,
      tier,
      payload,
      status:      'pending',
      createdAt:   new Date().toISOString(),
    };

    const item: ApprovalQueueItem = {
      id:          uuidv4(),
      action,
      requestedBy: 'AI Assist Panel',
      createdAt:   new Date().toISOString(),
      status:      'pending',
    };

    this.queue.set(item.id, item);
    return action;
  }

  getApprovalQueue(): ApprovalQueueItem[] {
    return Array.from(this.queue.values())
      .sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());
  }

  approve(id: string, decidedBy: string, note?: string): ApprovalQueueItem {
    return this.decide(id, 'approved', decidedBy, note);
  }

  reject(id: string, decidedBy: string, note?: string): ApprovalQueueItem {
    return this.decide(id, 'rejected', decidedBy, note);
  }

  async executeApproved(id: string): Promise<ApprovalQueueItem> {
    const item = this.queue.get(id);
    if (!item)                      throw new Error(`Approval item '${id}' not found`);
    if (item.status !== 'approved') throw new Error(`Action '${id}' has not been approved`);
    if (item.action.tier === 'executor') {
      // Only executor-tier actions can be auto-dispatched to the queue
      const payload = item.action.payload as { jobType?: string; data?: unknown };
      QueueEngine.enqueue(payload.jobType ?? 'GENERIC', payload.data ?? {}, { priority: 'high' });
    }
    return item;
  }

  private decide(id: string, status: ApprovalStatus, decidedBy: string, note?: string): ApprovalQueueItem {
    const item = this.queue.get(id);
    if (!item) throw new Error(`Approval item '${id}' not found`);
    const updated: ApprovalQueueItem = {
      ...item,
      status,
      decidedBy,
      decidedAt: new Date().toISOString(),
      note,
      action: { ...item.action, status },
    };
    this.queue.set(id, updated);
    return updated;
  }
}

export const AgenticService: IAgenticService = new AgenticServiceImpl();
